README.md
timeline-vue-component
==========================

[![Travis](https://img.shields.io/travis/ybg555/timeline-vue-component.svg)](https://travis-ci.org/ybg555/timeline-vue-component) [![npm](https://img.shields.io/npm/dm/timeline-vue-component.svg)](https://www.npmjs.com/package/timeline-vue-component) [![npm](https://img.shields.io/npm/v/timeline-vue-component.svg)](https://www.npmjs.com/package/timeline-vue-component)

时间轴组件

![](https://user-images.githubusercontent.com/6490371/38771016-1649e1b4-404e-11e8-9d8f-83e3c2c84fc6.png)

### Features

---

* 支持懒加载和分页
* 数据来源支持前端纯静态、异步接口获取
* slot 定制内容显示
* 支持布局和样式定制


### Installation

---

```shell
npm install timeline-vue-component --save
```


### Usages

---

```js
import Timeline from 'timeline-vue-component';

Vue.component('timeline', Timeline);
```

```html
<!--数据来源为前端静态数据-->

<timeline
  paginationType="frontend"
  :timelineData="data"
  :pageSize="10"
  :spacing="180">
  <!--开始图标内定制符号，选传-->
  <template slot="startCircle">
    <i class="el-icon-arrow-up"></i>
  </template>
  <!--结尾图标内定制符号，选传-->
  <template slot="endCircle">
    <i class="el-icon-arrow-down"></i>
  </template>
  <!--内容按需定制布局和数据排版，必传-->
  <template slot="list" scope="scope">
    {{ scope.data }}
  </template>
</timeline>
```

```html
<!--数据来源为异步接口数据-->
<!--组件内点下一步触发事件，onFetchRemoteData回调函数第一个参数 page => {current,size}，同步父组件分页值-->

<timeline
  paginationType="remote"
  :timelineData="data"
  :pageSize="10"
  @fetch-remote="onFetchRemoteData"
  :spacing="180">
  <!--开始图标内定制符号，选传-->
  <template slot="startCircle">
    <i class="el-icon-arrow-up"></i>
  </template>
  <!--结尾图标内定制符号，选传-->
  <template slot="endCircle">
    <i class="el-icon-arrow-down"></i>
  </template>
  <!--内容按需定制布局和数据排版，必传-->
  <template slot="list" scope="scope">
    {{ scope.data }}
  </template>
</timeline>
```

时间轴数据格式
```js
// :timelineData="data"
data: [
  { // 通过 scope 接收
    groupTile: '分组标题',
    title: '标题',
    content: [
      {
        field: '自定义内容',
      }
    ]
  },
]
```

### API

---

##### config

参数名 | 类型 | 默认值 | 备注
------------ | ------------- | ------------ | ---------
paginationType | String  | - | 数据来源类型 frontend、remote（必填）
timelineData | Array  | - | 时间轴数据（必填）
pageSize | Number  | - | 分页 （选填）
spacing | Number  | - | 时间轴左边和右边的间距 （选填）
titleWidth | Number  | - | 标题的宽度 （选填）
circleWidth | Number  | - | 中间节点（时间轴上的圆点）宽高 （选填）
startCircleWidth | Number  | - | 时间轴上开始和结束节点的宽高 （选填）


### Issues

---

Submit the [issues](https://github.com/ybg555/timeline-vue-component/issues) if you find any bug or have any suggestion.


### Contribution

---

Fork the [repository](https://github.com/ybg555/timeline-vue-component) and submit pull requests.


### Release Notes

---

see [CHANGELOG](https://github.com/ybg555/timeline-vue-component/blob/master/CHANGELOG.md)


### License

---

[![npm](https://img.shields.io/npm/l/timeline-vue-component.svg)]()



CHANGELOG.md
2.0.0 / 2018-04-30
==================

* slot 传递数据从 item.content 变成了 item，方便使用者获取整行数据。
* 增加数据来源paginationType [frontend, remote]

1.0.0 / 2018-04-14
==================

* init

vm.js
export default {
  name: 'Timeline',
  props: {
    timelineData: {
      type: Array,
      default() {
        return [];
      }
    },
    pageSize: {
      type: Number,
      default: 10
    },
    paginationType: {
      type: String,
      default: 'frontend'
    },
    spacing: {
      type: Number,
      default: 180
    },
    titleWidth: {
      type: Number,
      default: 100
    },
    circleWidth: {
      type: Number,
      default: 10
    },
    startCircleWidth: {
      type: Number,
      default: 30
    }
  },
  data() {
    return {
      timelineLoading: false,
      list: [],
      current: 1,
      next: true,
      remoteEmit: false,
    };
  },
  computed: {
    circleLeft() {
      return (this.titleWidth - this.spacing) / 2;
    },
    verticalBarLeft() {
      return (this.spacing + this.circleLeft) + this.circleWidth / 2 - 1;
    },
    startCircleLeft() {
      return (this.spacing - this.verticalBarLeft) + this.startCircleWidth / 2 - 1;
    }
  },
  watch: {
    timelineData: {
      handler(newlval) {
        if (this.paginationType === 'remote') {
          if (this.remoteEmit) {
            this._remoteRender(newlval);
          } else {
            console.info('异步从接口获取数据，请重置父组件中的请求参数: pagination.current = 1');
            this.list = [];
            this.current = 1;
            this._remoteRender(newlval);
          }
        }
        this.remoteEmit = false;
      },
      deep: true
    }
  },
  methods: {
    // 前端静态数据判断next
    _hasNext() {
      return this.timelineData.length > this.list.length;
    },
    // 前端静态数据init
    _appendList(timelineList) {
      if (this.paginationType !== 'remote') {
        timelineList.forEach((item, index) => {
          if (index < this.pageSize) {
            this.list.push(item);
          }
        });
        this.next = this._hasNext();
      }
    },
    // 前端静态数据分页
    _frontendPagination(index) {
      const lens = this.pageSize + index + 1;
      for (let i = index + 1; i < lens; i++) {
        if (this.timelineData[i]) {
          this.list.push(this.timelineData[i]);
        }
      }
      this.next = this._hasNext();
    },
    // remote 分页
    _remotePagination() {
      this.$emit('fetch-remote', {
        current: this.current,
        size: this.pageSize
      });
      this.remoteEmit = true;
    },
    _remoteRender(newlval) {
      newlval.forEach((item, index) => {
        if (index < this.pageSize) {
          this.list.push(item);
        }
      });
      if (newlval && newlval.length < Number(this.pageSize)) {
        this.next = false;
      } else {
        this.next = true;
      }
      this.current += 1;
    },
    onPagination(index) {
      if (this.next) {
        if (this.paginationType === 'remote') {
          this._remotePagination();
        } else {
          this._frontendPagination(index);
        }
      }
    }
  },
  created() {
    this._appendList(this.timelineData);
  }
};